### ISSUE TITLE
Missing packages installed by addon's install blueprint

### ISSUE THREAD

**Issue by san650** [tomdale/ember-cli-addon-tests#27]
I have an addon that installs NPM packages using the addon's install blueprint.

e.g.

`/blueprints/<addon-name>/index.js`

```js
/*jshint node:true*/
module.exports = {
  description: '',
  normalizeEntityName: function() {},

  afterInstall: function() {
    return this.addPackageToProject('ceibo');
  }
};
```

I'm using `ceibo` here but it could be any other NPM package.

The problem is that this package is not present when I create more than one application using `ember-cli-addon-tests`.

The package is defined in the `package.json` of the first app created by `ember-cli-addon-tests` but is not present in the `package.json` of the following apps created by `ember-cli-addon-tests`.

e.g.

```js
it('first test', function() {
  app = new AddonTestApp();
  return app.create('first-app').then(function() {
    assert.ok(fs.readFileSync(app.path('package.json').indexOf('ceibo') > -1)); // runs OK
  });
});

it('second test', function() {
  app = new AddonTestApp();
  return app.create('second-app').then(function() {
    assert.ok(fs.readFileSync(app.path('package.json').indexOf('ceibo') > -1)); // FAILS
  });
});
```

I've created a project with the minimum code to reproduce the issue [here](https://github.com/san650/add-package-to-project-issue).

You can see it fails [here](https://travis-ci.org/san650/add-package-to-project-issue/jobs/185308270).

**Comment by simonihmig** [comment]
Hi @san650, some time ago I added the support for the addon's default blueprint. Maybe I missed something there, will have a look! 

In the meantime you may want to run your tests with setting the DEBUG environment var to "ember-cli-addon-tests" like this:

    DEBUG=ember-cli-addon-tests mocha --recursive test/ 

This will give you some detailed debug messages about what is happening...

**Comment by san650** [comment]
@simonihmig Thanks for your response. I've run the tests using the DEBUG flag, here's the output

<details>
<summary>Output</summary>

```
$ DEBUG=ember-cli-addon-tests npm test

> add-package-to-project-issue@1.0.0 test /Projects/add-package-to-project-issue
> mocha --recursive test/

  01-test
  ember-cli-addon-tests created tmp; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg +0ms
  ember-cli-addon-tests no precooked node_modules +1ms
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg +1ms
  ember-cli-addon-tests no node_modules or bower_components +1ms
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine +0ms
  ember-cli-addon-tests running command=/Projects/add-package-to-project-issue/node_modules/ember-cli/bin/ember; args=new,--disable-analytics,--watcher=node,--skip-git,dummy; cwd=/private/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine +4ms
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine/dummy +2m
  ember-cli-addon-tests installing ember-data emberjs/data#master +0ms
  ember-cli-addon-tests installing ember-disable-prototype-extensions +3ms
  ember-cli-addon-tests installed ember-disable-prototype-extension +8s
  ember-cli-addon-tests installed ember-data emberjs/data#master +13s
  ember-cli-addon-tests found addon package.json; path=/Projects/add-package-to-project-issue/package.json +3ms
  ember-cli-addon-tests found addon package.json; path=/Projects/add-package-to-project-issue/package.json +2ms
  ember-cli-addon-tests symlinking add-package-to-project-issue +0ms
  ember-cli-addon-tests found addon package.json; path=/Projects/add-package-to-project-issue/package.json +5ms
  ember-cli-addon-tests installing add-package-to-project-issue as addon for application +0ms
  ember-cli-addon-tests installed ember canary +719ms
  ember-cli-addon-tests found addon package.json; path=/Projects/add-package-to-project-issue/package.json +3ms
  ember-cli-addon-tests running add-package-to-project-issue generator +2ms
  ember-cli-addon-tests running command=/Projects/add-package-to-project-issue/node_modules/ember-cli/bin/ember; args=generate,add-package-to-project-issue; cwd=/private/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine/dummy +0ms
  ember-cli-addon-tests DEPRECATION: Node v5.10.1 is no longer supported by Ember CLI. We recommend that you use the most-recent "Active LTS" version of Node.js.
 +830ms
  ember-cli-addon-tests WARNING: Node v5.10.1 is not tested against Ember CLI on your platform. We recommend that you use the most-recent "Active LTS" version of Node.js.
 +1ms
  ember-cli-addon-tests installing add-package-to-project-issue
 +47ms
  ember-cli-addon-tests   install addon ember-cli-mirage
 +8ms
  ember-cli-addon-tests Installed packages for tooling via npm.
 +16s
  ember-cli-addon-tests installing ember-cli-mirage
 +166ms
  ember-cli-addon-tests   create /mirage/config.js
 +13ms
  ember-cli-addon-tests   create /mirage/scenarios/default.js
 +0ms
  ember-cli-addon-tests   create /mirage/serializers/application.js
 +0ms
  ember-cli-addon-tests Installed addon package.
 +16ms
  ember-cli-addon-tests copying pristine app; from=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine/dummy; to=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/under-test/dummy +38ms
  ember-cli-addon-tests copying complete +32ms
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/under-test/dummy +0ms
  ember-cli-addon-tests chdir; path=/Projects/add-package-to-project-issue +1ms
  ember-cli-addon-tests copying fixtures; from=/Projects/add-package-to-project-issue/test/fixtures/dummy; to=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/under-test/dummy +0ms
    ✓ foo test
    ✓ bar test

  02-test
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg +14ms
  ember-cli-addon-tests existing node_modules and bower_components +0ms
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine +0ms
  ember-cli-addon-tests running command=/Projects/add-package-to-project-issue/node_modules/ember-cli/bin/ember; args=new,--skip-bower,--skip-npm,--disable-analytics,--watcher=node,--skip-git,second; cwd=/private/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine +1ms
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine/second +530ms
  ember-cli-addon-tests installing ember-data emberjs/data#master +1ms
  ember-cli-addon-tests found addon package.json; path=/Projects/add-package-to-project-issue/package.json +2ms
  ember-cli-addon-tests installing add-package-to-project-issue as addon for application +0ms
  ember-cli-addon-tests found addon package.json; path=/Projects/add-package-to-project-issue/package.json +2ms
  ember-cli-addon-tests running add-package-to-project-issue generator +2ms
  ember-cli-addon-tests running command=/Projects/add-package-to-project-issue/node_modules/ember-cli/bin/ember; args=generate,add-package-to-project-issue; cwd=/private/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine/second +0ms
  ember-cli-addon-tests
Missing bower packages:
Package: ember
  * Specified: canary
  * Installed: (not installed)

Package: ember-cli-shims
  * Specified: 0.1.3
  * Installed: (not installed)

Run `bower install` to install missing dependencies.

 +503ms
  ember-cli-addon-tests copying pristine app; from=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/pristine/second; to=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/under-test/second +6ms
  ember-cli-addon-tests copying complete +30ms
  ember-cli-addon-tests chdir; path=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/under-test/second +0ms
  ember-cli-addon-tests chdir; path=/Projects/add-package-to-project-issue +0ms
  ember-cli-addon-tests copying fixtures; from=/Projects/add-package-to-project-issue/test/fixtures/second; to=/var/folders/j4/g0k_9zns07v0ptp8l0dw117h0000gn/T/d-1161120-51636-1ss3ojg/under-test/second +1ms
    1) foo test

    2) bar test



  2 passing (3m)
  2 failing

  1) 02-test foo test:

      AssertionError: "ember-cli-mirage" not found in package.json
      + expected - actual

      -false
      +true

      at assertPackage (test/02-test.js:12:12)
      at Context.<anonymous> (test/02-test.js:22:5)

  2) 02-test bar test:

      AssertionError: "ember-cli-mirage" not found in package.json
      + expected - actual

      -false
      +true

      at assertPackage (test/02-test.js:12:12)
      at Context.<anonymous> (test/02-test.js:26:5)



npm ERR! Test failed.  See above for more details.
```
</details>

Do you see anything strange on the output?

**Comment by simonihmig** [comment]
@san650 I think I found something regarding this issue. 

This addon caches `node_modules` and `bower_components` for subsequent runs. That does not seem to work when changing the app name, which you do here: https://github.com/san650/add-package-to-project-issue/blob/master/test/02-test.js#L18

This can be seen here:
```
ember-cli-addon-tests existing node_modules and bower_components +0ms
  ember-cli-addon-tests chdir; path=/var/folders/q3/yzggf_l965zbzps_bk8xr37c0000gn/T/d-11703-15846-cjqa8f.y1g79wl8fr/pristine +0ms
  ember-cli-addon-tests running command=/Users/simonihmig/Projects/add-package-to-project-issue/node_modules/ember-cli/bin/ember; args=new,--skip-bower,--skip-npm,--disable-analytics,--watcher=node,--skip-git,second; cwd=/private/var/folders/q3/yzggf_l965zbzps_bk8xr37c0000gn/T/d-11703-15846-cjqa8f.y1g79wl8fr/pristine +4ms
```

It assumes existing `node_modules` and `bower_components` (as they *are* present for the 'dummy' app), so it runs `ember new` with the `--skip-bower,--skip-npm` flags. However it creates the app under the new name 'second', so it has no `node_modules` and `bower_components` there.

The `ember generate` command then fails as you would expect given the missing deps:
```
  ember-cli-addon-tests running command=/Users/simonihmig/Projects/add-package-to-project-issue/node_modules/ember-cli/bin/ember; args=generate,add-package-to-project-issue; cwd=/private/var/folders/q3/yzggf_l965zbzps_bk8xr37c0000gn/T/d-11703-15846-cjqa8f.y1g79wl8fr/pristine/second +0ms
  ember-cli-addon-tests 
Missing bower packages: 
Package: ember
  * Specified: canary
  * Installed: (not installed)
...
```

So this is not related to the default blueprint, any ember-cli command would fail.

I am not sure if this *should* work. @tomdale? However for you the fix is quite easy, just change the second test to also use the 'dummy' app name. Just did that and it works then! 

I think you don't have to use different app names, do you? Things like the test fixtures will not be reused between separate runs (different `AddonTestApp.create` calls)

**Comment by san650** [comment]
Thanks @simonihmig for investigating the issue.

I'm using different app names so I can use different fixture folders for each test -- As I understand, the fixture folder is derived from the app name.

I'm already doing this on the test suite of an addon (see https://github.com/san650/ember-web-app/tree/master/node-tests/acceptance/fixtures) to simulate different applications states... e.g. one app with X configuration and files and another with Y configuration and files. Then in the acceptance test I create the app using the different fixture folders (see https://github.com/san650/ember-web-app/blob/master/node-tests/acceptance/manifest-test.js).

A workaround I found for now is to update the dev packages of the generated app to include the missing dependency (see https://github.com/san650/ember-web-app/blob/fcb8af79472703c3439618059716549a126f1dca/node-tests/acceptance/manifest-test.js#L48-L55).

Looking at the fastboot tests it seems they are using multiple apps in their acceptance tests to test different conditions. See:

* https://github.com/ember-fastboot/ember-cli-fastboot/tree/master/test/fixtures
* https://github.com/ember-fastboot/ember-cli-fastboot/blob/master/test/asset-rewriting-test.js#L15
* https://github.com/ember-fastboot/ember-cli-fastboot/blob/master/test/fastboot-location-test.js#L14

About the failing example I linked in my first comment, it fails when the addon under test has a default blueprint which installs NPM packages (or ember addons). If the addon doesn't install any extra NPM package then the tests work just fine.

So, my guess is that is related to the fact that the cached node_modules folder does not contain the NPM packages installed through the default blueprint of the addon.

**Comment by simonihmig** [comment]
> I'm using different app names so I can use different fixture folders for each test -- As I understand, the fixture folder is derived from the app name.

You can use `fixturesPath`option: https://github.com/tomdale/ember-cli-addon-tests#options

> About the failing example I linked in my first comment, it fails when the addon under test has a default blueprint which installs NPM packages (or ember addons). If the addon doesn't install any extra NPM package then the tests work just fine.

You are right, I misinterpreted the debug statements there. Sorry for the confusion! So having different app names should be working...

> So, my guess is that is related to the fact that the cached node_modules folder does not contain the NPM packages installed through the default blueprint of the addon.

Actually it does contain it, but the `generate` command is run before the cached node_modules is symlinked to the new ('second') app. So it *is* related to the default blueprint, and therefore it's me who is to blame for that! 😳

Have a fix for this already, will just have to test it a little bit more...

**Comment by san650** [comment]
Great!

> Have a fix for this already, will just have to test it a little bit more...

Let me know if you want me to try it out.

**Comment by simonihmig** [comment]
Sure, here we go: just pushed the fix to this branch: https://github.com/tomdale/ember-cli-addon-tests/tree/fix-generate-command-order. Please check this before getting this merged. Put `tomdale/ember-cli-addon-tests#fix-generate-command-order` in your package.json for the `ember-cli-addon-tests` dependency.

**Comment by san650** [comment]
@simonihmig I tested it and it fixes the issue 😄

**Comment by simonihmig** [comment]
Great to hear. Will merge and create a patch release today!

**Comment by simonihmig** [comment]
Sorry for the delay. Just published the fix as v0.4.2!
